### PR TITLE
Wiki deployment script with new wiki folder

### PR DESCRIPTION
**Pull request by hadarai**
I've added script that deploys documents from folder `/wiki` to our GitHub Wiki.
And a first version of that wiki with copies of README.
To start working script needs addition of correct secret value in repo's settings.

Script uses this plugin: https://github.com/marketplace/actions/github-wiki-action
About secrets you can read here: https://docs.github.com/en/actions/reference/encrypted-secrets
Working example of wiki deployment: https://github.com/hadarai/RPIS_zad_dodatowe

### CODE REVIEW

File: wiki/Debugging.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# Debugging
+
+Custom debugging tools for GDB.
+
+Running
+---
+
+After you start the kernel in debugging mode, i.e. using
+[launch](https://github.com/cahirwpz/mimiker#running) script with `-d` flag,
+you can use following custom commands:
```
Comment (pj1031999) [inline wiki/Debugging.md:10]: IIRC `-d` is required now.

File: wiki/Compiler-toolchain.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+On Debian, executing the following command should suffice:
+
+```
+$ sudo apt-get install automake bison ca-certificates curl flex gcc g++ gettext lhasa libtool make patch pkg-config python3 python3-setuptools quilt texinfo xz-utils zip libpython3-dev debhelper
+```
+
+### Configuration
+
+You can change the default configuration by editing the file `config.mk`.
+
```
Comment (pj1031999) [inline wiki/Compiler-toolchain.md:17]: How `config.mk` is related to toolchain?

File: wiki/Debugging.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+`ktrace` parameter selects on which predefined function create/remove
+breakpoints on:
+
+* `thread-create` - function [thread_create](https://github.com/cahirwpz/mimiker/blob/master/sys/thread.c)
+* `ctx-switch` - function [ctx_switch](https://github.com/cahirwpz/mimiker/blob/master/mips/switch.S)
+
+How to debug user programs?
+---
+
+According to [mimiker.ld](https://github.com/cahirwpz/mimiker/blob/master/user/libmimiker/extra/mimiker.ld)
```
Comment (pj1031999) [inline wiki/Debugging.md:35]: Dead link.

File: wiki/Debugging.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+* `klog` - all log messages currently saved in the kernel,
+* `threads` - all existing threads,
+* `tlb` - Translation Lookaside Buffer with addresses and flags marking if page
+  is dirty (D) and/or global (G).
+
+`ktrace` parameter selects on which predefined function create/remove
+breakpoints on:
+
+* `thread-create` - function [thread_create](https://github.com/cahirwpz/mimiker/blob/master/sys/thread.c)
+* `ctx-switch` - function [ctx_switch](https://github.com/cahirwpz/mimiker/blob/master/mips/switch.S)
```
Comment (pj1031999) [inline wiki/Debugging.md:30]: Dead links.

File: wiki/Debugging.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+`kdump` parameter select which state to dump:
+
+* `free_pages` - list of free pages per segment with virtual and physical
+   addresses,
+* `segments` - list all memory segments, incl. start, end addresses and
+   number of pages (currently just one),
+* `klog` - all log messages currently saved in the kernel,
+* `threads` - all existing threads,
+* `tlb` - Translation Lookaside Buffer with addresses and flags marking if page
+  is dirty (D) and/or global (G).
```
Comment (pj1031999) [inline wiki/Debugging.md:24]: It's true only for mips.

File: wiki/Debugging.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+user space binaries are loaded at *0x400000* address. You could stop an entry
+point to the program listed in ELF header. However using **gdb** without
+debugging information is not practical in long-term. So... how to stop debugger
+at user program `main` procedure and step through it in casual way?
+
+You have to issuing following command (in this case for `utest` program) within
+**gdb** session:
+
+> `add-symbol-file user/utest/utest.uelf 0x400000`
+
```
Comment (pj1031999) [inline wiki/Debugging.md:45]: I'm not sure if it 's a correct path.

File: wiki/Github-issue-PR-labels.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+---
+
+ - ![#a020f0](https://placehold.it/15/a020f0/000000?text=+) `mockup` - (usually applied to pull requests) This branch is not supposed to be merged, and only provides a demonstration of a feature or an interface.
+ - ![#a020f0](https://placehold.it/15/a020f0/000000?text=+) `research` - This task doesn't require writing source code, but investigating solutions or tools available.
+ - ![#cc317c](https://placehold.it/15/cc317c/000000?text=+) `question` - This entry is used for discussion, not actual development.
+ - ![#bfe5bf](https://placehold.it/15/bfe5bf/000000?text=+) `proposal` - We have not decided yet (or the author does not know) whether the branch is worth merging, or whether the issue is valid.
+ - ![#bfd4f2](https://placehold.it/15/bfd4f2/000000?text=+) `wish` - This task is of low priority, and may have a very vague description.
+ - ![#ffffff](https://placehold.it/15/ffffff/000000?text=+) `wontfix` - We've decided not to fix the bug, either because it is invalid, unfixable, or fixing it does not follow our general agenda.
```
Comment (pj1031999) [inline wiki/Github-issue-PR-labels.md:25]: We have more labels.

File: wiki/Home.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+5. [Separation of mechanism and policy](https://en.wikipedia.org/wiki/Separation_of_mechanism_and_policy)
+6. [You aren't gonna need it](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)
+7. [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter)
+
+### Useful links:
+
+* [L4 Kernel Reference Manual](http://www.l4ka.org/l4ka/l4-x2-r7.pdf) Version X.2
+* [Linux Kernel Cross Reference](http://lxr.free-electrons.com/)
+* [FreeBSD Kernel Cross Reference](http://fxr.watson.org/fxr/source/)
+* [Minix 3.3.0 Cross Reference](http://users.sosdg.org/~qiyong/mxr/source/minix/kernel/)
```
Comment (pj1031999) [inline wiki/Home.md:25]: We mostly use bxr.su.

File: wiki/Related-scripts.md [hunk 9]
```diff
@@ -0,0 +1,2 @@
+Mimiker-related scripts
+---
```
Comment (pj1031999) [inline wiki/Related-scripts.md:2]: Should it be empty?

File: wiki/Running-DOOM-on-mimiker.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+```
+git clone https://github.com/rafalcieslak/headless_doom user/headless_doom
+```
+
+* Place your `doom.wad` inside `./user/headless_doom/`. This has to be the `doom.wad` from *The Ultimate Doom 1.10*, no other IWADs are supported. The `Makefile` for headless_doom will check the md5 sum of the wad file. Note that the file name is case-sensitive. [*If you don't own Ultimate DOOM, you can get it on [Steam](http://store.steampowered.com/app/2280/) or [GOG](https://www.gog.com/game/the_ultimate_doom). It frequently goes on sale for under 3€. If you are a mimiker developer, you can find `doom.wad` on the mimiker dev server.*]
+
+```
+cp /path/to/your/doom/doom.wad user/headless_doom/.
+```
```
Comment (pj1031999) [inline wiki/Running-DOOM-on-mimiker.md:13]: Is it up to date?

File: wiki/Test-infrastructure.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+### Running tests
+
+If you just want to run all available tests please use `./run_tests.py`. This is
+the command that is used by continuous integration. Some useful arguments are
+listed below:
+
+* `--infinite` - keep testing until some error is found.
+* `--non-interactive` - do not run interactive gdb session if tests fail.
+* `--thorough` - generate much more test seeds. Testing will take much more time.
```
Comment (pj1031999) [inline wiki/Test-infrastructure.md:12]: `--board` parameter is also very important.

File: wiki/Test-infrastructure.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+* `repeat=UINT` - Specifies the number of (shuffled) repetitions of each test
+  when using `test=all`.
+  When running user-specified tests (i.e. when `test` is not equal to `all`),
+  `repeat` specifies the number of times each test will be run. For example,
+  `test=test1,test2 repeat=4` will run `test1` 4 times and then run `test2` 4 times.
+
+### Programming tests
+
+##### Kernel tests
+Located in `$(TOPDIR)/tests`.
```
Comment (pj1031999) [inline wiki/Test-infrastructure.md:37]: Is `$(TOPDIR)` equal to `/sys`?

File: wiki/newcomers-FAQ.md [hunk 13]
```diff
@@ -0,0 +1 @@
+# Newcomers FAQ
```
Comment (pj1031999) [inline wiki/newcomers-FAQ.md:1]: Should it be empty?

File: wiki/Test-infrastructure.md [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+Macros to register tests:
+* `KTEST_ADD(name, func, flags)`
+* `KTEST_ADD_RANDINT(name, func, flags, max)` - need to cast function pointer to
+  `(int (*)(void))`
+
+Where `name` is test name, `func` is pointer to test function,
+flags as mentioned below, and `max` is maximum random argument fed to the test.
+
+##### User tests
+Located in `$(TOPDIR)/user/utest`.
```
Comment (pj1031999) [inline wiki/Test-infrastructure.md:50]: `/bin/utest`

**Review by pj1031999 — COMMENTED** [review]
What is .DB_Store?

File: wiki/Coding-conventions.md [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+**Explicit size for basic data types.**
+
+> Please use data types from [stdint.h](http://www.cplusplus.com/reference/cstdint/), namely:
+> * `uint8_t` instead of `unsigned char`
+> * `uint16_t` instead of `unsigned short int`
+> * `int16_t` instead of `short int`
+> * `int32_t` instead of `int`
+> * `uint32_t` instead of `unsigned int`
+> * `int64_t` instead of `long`
+> * `uint64_t` instead of `unsigned long`
```
[the remark this answers is not in the document]
Comment (hadarai) [reply wiki/Coding-conventions.md:16]: It's copied file from current wiki 😄

File: wiki/newcomers-FAQ.md (hunk 13 above)
Comment (hadarai) [reply wiki/newcomers-FAQ.md:1]: It's a proposition that such file should exist, but we yet have to discuss it. I'll delete it for now.

File: wiki/Test-infrastructure.md (hunk 12 above)
Comment (hadarai) [reply wiki/Test-infrastructure.md:37]: No idea, but `TOPDIR` doesn't look like it, so I'll change it.

File: wiki/_Sidebar.md [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+- [Github issue PR labels](https://github.com/hadarai/mimiker-wiki/wiki/Github-issue-PR-labels)
+
+**Developing documents**
+- [Test infrastructure](https://github.com/hadarai/mimiker-wiki/wiki/Test-infrastructure)
+- [Debugging](https://github.com/hadarai/mimiker-wiki/wiki/Debugging)
+- [Libkern](https://github.com/hadarai/mimiker-wiki/wiki/Libkern-information)
+
+**Miscellaneous**
+- [Related scripts](https://github.com/hadarai/mimiker-wiki/wiki/Related-scripts)
+- [Running Doom](https://github.com/hadarai/mimiker-wiki/wiki/Running-DOOM-on-mimiker)
```
Comment (pj1031999) [inline wiki/_Sidebar.md:20]: Why wiki is located in https://github.com/hadarai instaed of https://github.com/cahirwpz?

Comment (hadarai) [reply wiki/_Sidebar.md]: That's a very good point.

File: wiki/Test-infrastructure.md (hunk 11 above)
Comment (hadarai) [reply wiki/Test-infrastructure.md:12]: It needs proper description to be added

**Comment by hadarai** [comment]
The decision was made that in this commit there will be only `Home.md` page of wiki. Rest will soon follow, when updated.

**Review by pj1031999 — CHANGES_REQUESTED** [review]
Please remove unused functionality, check licence of copied files or rewrite them (from 0) for our needs.

File: wiki/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,10 @@
+# This file is a fork of GitHub Wiki Action plugin by Andrew Chen Wang
+# https://github.com/Andrew-Chen-Wang/github-wiki-action
+
+FROM alpine/git
+
+RUN apk add rsync
+
+COPY entrypoint.sh /entrypoint.sh
+RUN chmod +x /entrypoint.sh
+ENTRYPOINT ["/entrypoint.sh"]
```
[the remark this answers is not in the document]
Comment (hadarai) [reply wiki/Dockerfile:10]: Licence is Apache Licence 2.0
I spoke with Cahir about using his code and adding information about source of the code seems to be enough

File: wiki/entrypoint.sh [hunk 18]
```diff
@@ -0,0 +1,75 @@
+#!/bin/sh
+
+# This file is a fork of GitHub Wiki Action plugin by Andrew Chen Wang
+# https://github.com/Andrew-Chen-Wang/github-wiki-action
+
+TEMP_CLONE_FOLDER="temp_wiki_$GITHUB_SHA"
+TEMP_EXCLUDED_FILE="temp_wiki_excluded_$GITHUB_SHA.txt"
+
+if [ -z "$GH_TOKEN" ]; then
+  echo "GH_TOKEN ENV is missing. Use $\{{ secrets.GITHUB_TOKEN }} or a PAT if your wiki repo is different from your current repo."
```
[the remark this answers is not in the document]
Comment (hadarai) [reply wiki/entrypoint.sh:10]: Private Access Token. Token used to validate push to private repository

File: wiki/entrypoint.sh [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+    rm -r $TEMP_CLONE_FOLDER/$file
+  done
+fi
+
+echo "Pushing to Wiki"
+cd $TEMP_CLONE_FOLDER
+git add .
+git commit -m "$message"
+git push --set-upstream https://$GH_NAME:$GH_TOKEN@github.com/$REPO.wiki.git master || git push --set-upstream https://$GH_TOKEN@github.com/$REPO.wiki.git master
+
```
[the remark this answers is not in the document]
Comment (hadarai) [reply wiki/entrypoint.sh:75]: That's right, I'll check both and pick safer one

File: wiki/entrypoint.sh [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+else
+  for file in $EXCLUDED_FILES; do
+    echo "$file" >> ./$TEMP_EXCLUDED_FILE
+  done
+  rsync -av --delete $WIKI_DIR $TEMP_CLONE_FOLDER/ --exclude .git --exclude-from=$TEMP_EXCLUDED_FILE
+  # Delete files in target repo if it was a reminant.
+  for file in $EXCLUDED_FILES; do
+    rm -r $TEMP_CLONE_FOLDER/$file
+  done
+fi
```
[the remark this answers is not in the document]
Comment (hadarai) [reply wiki/entrypoint.sh:68]: We may like to have it in the future... but i guess we'll add it then.

File: wiki/entrypoint.sh [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+
+if [ -z "$GH_NAME" ]; then
+  echo "GH_NAME ENV is missing. Use the username for a user that can push to this repository."
+  exit 1
+fi
+
+if [ -z "$REPO" ]; then
+  echo "REPO ENV is missing. Using the current one."
+  REPO=$GITHUB_REPOSITORY
+fi
```
[the remark this answers is not in the document]
Comment (hadarai) [reply wiki/entrypoint.sh]: I guess we don't. removed

File: wiki/entrypoint.sh [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+
+git pull https://$GH_TOKEN@github.com/$REPO.wiki.git
+cd ..
+
+# Get commit message
+if [ -z "$WIKI_PUSH_MESSAGE" ]; then
+  message=$(git log -1 --format=%B)
+else
+  message=$WIKI_PUSH_MESSAGE
+fi
```
[the remark this answers is not in the document]
Comment (hadarai) [reply wiki/entrypoint.sh]: That's correct, removed

File: wiki/entrypoint.sh (hunk 18 above)
[the remark this answers is not in the document]
Comment (hadarai) [reply wiki/entrypoint.sh]: We won't need that here

File: wiki/action.yml [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+  WIKI_PUSH_MESSAGE:
+    description: 'The commit message for the wiki commit. Default is the commit to actual repository.'
+    required: false
+  EXCLUDED_FILES:
+    description: 'A list of files you want to exclude (separate with spaces).'
+    required: false
+
+branding:
+  icon: upload-cloud
+  color: blue
```
[the remark this answers is not in the document]
Comment (hadarai) [reply wiki/action.yml]: It's much more simple now, but I can't merge it with `wiki_sync.yml`. I can remove it only if we don't want to pass any variables.

File: wiki/entrypoint.sh [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+
+if [ -z "$REPO" ]; then
+  echo "REPO ENV is missing. Using the current one."
+  REPO=$GITHUB_REPOSITORY
+fi
+
+if [ -z "$WIKI_DIR" ]; then
+    echo "WIKI_FOLDER ENV is missing, using default wiki/"
+    WIKI_DIR='wiki/'
+fi
```
[the remark this answers is not in the document]
Comment (hadarai) [reply wiki/entrypoint.sh]: Okay, i did so.

File: .github/workflows/deploy_wiki.yml [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+
+jobs:
+  deploy-wiki:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout on container
+      uses: actions/checkout@v2 
+
+    - name: Push Wiki Changes
+      # here action.yml script is run
```
Comment (pj1031999) [inline .github/workflows/deploy_wiki.yml:22]: outdated comment

File: wiki/entrypoint.sh [hunk 26]
```diff
@@ -0,0 +1,53 @@
+#!/bin/sh
+
+# This file is a fork of GitHub Wiki Action plugin by Andrew Chen Wang
+# https://github.com/Andrew-Chen-Wang/github-wiki-action
+
+TEMP_CLONE_FOLDER="temp_wiki_$GITHUB_SHA"
+
+# This script assumes there is rsync and git installed in executing container.
```
Comment (pj1031999) [inline wiki/entrypoint.sh:8]: It's true for ubuntu20.04 [1] image which is used here.

[1] https://github.com/actions/virtual-environments/tree/main/images/linux

**Review by pj1031999 — APPROVED** [review]
Now I don't see a simpler solution. I left minor comments but overall experience is good.

LGTM

File: wiki/entrypoint.sh (hunk 26 above)
Comment (hadarai) [reply wiki/entrypoint.sh:8]: should I add that information to that comment?

Comment (pj1031999) [reply wiki/entrypoint.sh:8]: It's not necessary for me.

**Comment by pwit81** [comment]
LGTM

**Review by pwit81 — APPROVED** [review]

**Pull request merged**